pic-sure-auth-services/src/main/java/edu/harvard/hms/dbmi/avillach/auth/filter/JWTFilter.java

```diff
@@ -0,0 +1,223 @@
+
+import java.io.IOException;
+import java.util.Set;
+
+
+/**
+ * The main gate for PSAMA that filters all incoming requests against PSAMA.
+ * <h3>Design Logic</h3>
+ * <ul>
+ *     <li>All incoming requests pass through this filter.</li>
+ *     <li>To pass this filter, the incoming request needs a valid bearer token in its HTTP Authorization Header
+ *     to represent a valid identity behind the token. </li>
+ *     <li>In some cases, the incoming request doesn't need to hold a token. For example, when the request is to the <code>authentication</code>
+ *     endpoint, <code>swagger.json</code>, or <code>swagger.html</code>.</li>
+ * </ul>
+ */
+
+@Component
+@Order(1)
+public class JWTFilter extends OncePerRequestFilter {
+
+    private final static Logger logger = LoggerFactory.getLogger(JWTFilter.class);
+
+    private final TOSService tosService;
+
+    private final String userClaimId;
+
+    private final JWTUtil jwtUtil;
+    private final CustomUserDetailService customUserDetailService;
+
+    @Autowired
+    public JWTFilter(TOSService tosService,
+                     @Value("${application.user.id.claim}") String userClaimId, JWTUtil jwtUtil, CustomUserDetailService customUserDetailService) {
+        this.tosService = tosService;
+        this.userClaimId = userClaimId;
+        this.jwtUtil = jwtUtil;
+        this.customUserDetailService = customUserDetailService;
+    }
+
+    /**
+     * Filter implementation that performs authentication and authorization checks based on the provided request headers.
+     * The filter checks for the presence of the "Authorization" header and validates the token.
+     * It sets the appropriate security context based on the type of token (long term token or PSAMA application token) and
+     * performs the necessary checks to ensure that the user or application is authorized to access the requested resource.
+     * This filter is called by the configured security filter chain in the SecurityConfig class.
+     *
+     * @param request     the HttpServletRequest object
+     * @param response    the HttpServletResponse object
+     * @param filterChain the FilterChain object
+     * @throws IOException if an I/O error occurs during the execution of the filter
+     */
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, @NonNull HttpServletResponse response, @NonNull FilterChain filterChain) throws IOException, ServletException {
+        // Get headers from the request
+        String authorizationHeader = request.getHeader("Authorization");
+
+        if (!StringUtils.isNotBlank(authorizationHeader)) {
+            // If the header is not present, we allow the request to pass through
+            // without any authentication or authorization checks
+            filterChain.doFilter(request, response);
+        } else {
+            // If the header is present, we need to check the token
+            String token = authorizationHeader.substring(6).trim();
```

Comment:
One thing that could make this a bit easier to understand would be to put the string constant that you're omitting in the code.
```suggestion
            String token = authorizationHeader.substring("blah".length).trim();
```